lib/mongo/server/connection_pool.rb

```diff
@@ -377,7 +378,7 @@ def check_out(connection_global_id: nil)
                 #
                 # Ruby does not allow a thread to lock a mutex which it already
                 # holds.
-                if unsynchronized_size < max_size
+                if max_size == 0 || unsynchronized_size < max_size
```

Comment:
I am not a fan of max_size attribute of the pool being zero but the alternative I can think of (have max_size be nil for an unlimited pool) is also confusing: nil max_size in constructor argument is changed to max_size=20, max_size=0 in constructor argument is changed to max_size nil. Unless you have some other ideas we can leave this the way it is typed here for now.

Reply:
We could theoretically use `Float::INFINITY` for the `max_size`, but it is (obviously) float.